vllm/model_executor/guided_decoding/xgrammar_utils.py

```diff
@@ -0,0 +1,120 @@
+import re
+
+
+def grammar_is_likely_lark(grammar_str: str) -> bool:
+    """
+    Check if grammar appears to use Lark syntax.
+    
+    Args:
+        grammar_str: Input grammar string
+        
+    Returns:
+        bool: True if grammar appears to be in Lark format, False otherwise
+        
+    Examples:
+        >>> grammar_is_likely_lark("rule: 'abc'")
+        True
+        >>> grammar_is_likely_lark("rule ::= 'abc'")
+        False
+    """
+    if not grammar_str or not isinstance(grammar_str, str):
+        return False
+
+    for line in grammar_str.split('\n'):
+        # Remove both comment styles
+        line = re.sub(r'(#|//).*$', '', line).strip()
+        if not line:
+            continue
+
+        # Look for Lark-style rule definitions
+        if ':' in line and '::=' not in line:
+            return True
+
+        # Look for Lark-specific features
+        if any(pattern in line for pattern in ['?start:', '|', '~']):
+            return True
+
+    return False
+
+
+def convert_lark_to_gbnf(grammar_str: str) -> str:
+    """
```

Comment:
any sources or references that we can cite on this?

Reply:
Unfortunately I couldn't find an existing converter, so I made this myself (with the help of AI) using the [GBNF reference](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md) and the [Lark grammar reference](https://lark-parser.readthedocs.io/en/latest/grammar.html), so I can leave these links.

I tested that it works for our SQL-based grammars in examples and CI, but I wouldn't trust it blindly.